compiler/src/iree/compiler/Dialect/HAL/IR/HALAttrs.td

```diff
@@ -1011,4 +1011,67 @@ def HAL_DevicePromiseAttr : AttrDef<HAL_Dialect, "DevicePromise", [
   let hasCustomAssemblyFormat = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// #hal.device.topology<...>
+//===----------------------------------------------------------------------===//
+
+def HAL_DeviceTopologyAttr : AttrDef<HAL_Dialect, "DeviceTopology", [
+  DeclareAttrInterfaceMethods<Stream_AffinityTopologyAttr, [
+    "requiresTransfer",
+  ]>
+]> {
+  let mnemonic = "device.topology";
+  let summary = [{Device connection topology specification.}];
+  let description = [{
+    Specifies the topology of connections between devices, including
+    information about which devices can share data without transfers.
+
+    This attribute is attached to the top-level module as `stream.topology`
+    to inform Stream dialect passes about device connectivity.
+
+    Example:
+    ```mlir
+    module {
+      stream.topology = #hal.device.topology<
+        links = [
+          {[@device_a, @device_b], transfer_required = false},
+          {[@device_a, @device_c], transfer_required = true}
```

Comment:
Let's make the per-link entry be an attr dict instead - that way we can add more link properties in the future (like link weight, bandwidth, latency, etc). We could use `struct` to start but may eventually want it to be extensible for target-specific link information (such as between two GPU devices that share an SOC with some magic link, or infiniband or PCIE weighting, etc).

The syntax would just get some extra braces:
```mlir
{[@device_a, @device_b], {transfer_required = false}}
```